### PR TITLE
Update `deploy` command to build site before deployment

### LINKED ISSUE #534: markbind deploy: build before deploying
Current: `markbind deploy` does not build but deploy only.

In most cases users would want to deploy the latest version of the site. Should it build by default with possibly a `--no-build` option?

### PR DESCRIPTION
**Pull request by lhw-1**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [X] Feature addition or enhancement
- [ ] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

**Overview of changes:**

Resolves #534.

The `markbind deploy` command now builds and generates the site by default, meaning that authors no longer need to run `markbind build` before `markbind deploy` as a mandatory step.

This PR also introduces a `--no-build` flag to the `deploy` command, so that authors can choose to instead deploy from a site that was generated previously (i.e. the behavior of the `deploy` command before this PR). If the author has generated the site with specific arguments (e.g. `baseUrl`), this option gives them the flexibility to use the specific generation for deployment. 

The assumption is that the general use case for the two commands would be `markbind build` and `markbind deploy` without additional arguments; this PR simply consolidates the two into one, while preserving the functionality of the original `markbind deploy` if needed.

**Anything you'd like to highlight/discuss:**

As an extension to this PR, we can possibly explore improving the `deploy` command to "remember" previous choices, e.g. if the site was built with specific arguments, then the generation process in the next `markbind deploy` could re-use these choices. More considerations would be needed here to avoid inadvertently complicating the commands. We could also consider allowing the `deploy` command to take in the same arguments as the `build` command, but again, further considerations would be needed to take care in not polluting the functionality of the `deploy` command.

Additionally, I think that we should refactor the `cli/index.js` file in the near future (as mentioned in #1756 [comment](https://github.com/MarkBind/markbind/issues/1756#issuecomment-1466383919)) so as to encourage further modularity and easier development / debugging for the CLI commands and processes.

**Testing instructions:**

**Proposed commit message: (wrap lines at 72 characters)**
```
Update `deploy` command to build site before deployment

The `deploy` command deploys without first building the 
latest version of the site.

The burden is placed on the user to always remember to run 
the `build` command before deploying the new site.

Having the `deploy` command build the site by default allows 
users to build the site with new changes and deploy the site 
with a single command.

Let's incorporate the site generation process into the `deploy` 
command, while providing an optional `--no-build` flag to skip 
the site generation as needed.

Generating the site as part of the deployment is preferable 
as users will need to build before deployment in most cases.
```

---

**Checklist:** :ballot_box_with_check:

- [X] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [X] Linked all related issues
- [ ] No unrelated changes

### CODE REVIEW

File: docs/userGuide/addingPages.md [hunk 1]
```diff
@@ -68,7 +68,7 @@ The general approach is as follows:
 1. Create a folder with the path that you want the HTML web page to be available at.
 1. Exclude the files from being further processed by MarkBind by adding them to the <trigger trigger="click" for="modal:addingPages-prevent-processing">`pagesExclude` attribute of the `site.json` file</trigger>.
 1. Generate the HTML files and put them in that folder (This can be done automatically by your build tool, either locally, or as part of your continuous integration process).
-1. Run `markbind build` to generate your MarkBind site.
+1. (Optional) Run `markbind build` to generate your MarkBind site.
```
Comment (EltonGohJH) [inline docs/userGuide/addingPages.md:71]: If the user is not using deploy and they deploying it via some other way, they would need to build right?

File: docs/userGuide/deployingTheSite.md [hunk 2]
```diff
@@ -23,17 +23,28 @@
 
 1. Set the [`baseUrl` property of the `site.json` file](siteJsonFile.html#baseurl) to match the deploy location.
 1. (Optional) Use the [`markbind serve` command](cliCommands.html#serve-command) to stage the site locally and confirm the contents are as expected.
-1. Use the [`markbind build` command](cliCommands.html#build-command) to generate the site from source files. That command puts the generated site files in a directory named `_site` (you can change the output directory using parameters supplied to the command).
+1. (Optional) Use the [`markbind build` command](cliCommands.html#build-command) to generate the site from source files. That command puts the generated site files in a directory named `_site` (you can change the output directory using parameters supplied to the command). Note that the [`markbind deploy` command](cliCommands.html#deploy-command) generates the site from source files by default as well.
```
Comment (EltonGohJH) [inline docs/userGuide/deployingTheSite.md:26]: same here.

File: packages/cli/index.js [hunk 3]
```diff
@@ -98,7 +97,9 @@ program
   .option('-p, --port <port>', 'port for server to listen on (Default is 8080)')
   .option('-s, --site-config <file>', 'specify the site config file (default: site.json)')
   .option('-d, --dev', 'development mode, enabling live & hot reload for frontend source files.')
+  .description('build then serve a website from a directory')
   .action((userSpecifiedRoot, options) => {
+    logger.info(options);
```
Comment (EltonGohJH) [inline packages/cli/index.js:102]: If you find printing this useful, should we print this in other commands too?

File: docs/userGuide/addingPages.md (hunk 1 above)
Comment (lhw-1) [reply docs/userGuide/addingPages.md:71]: That's right! I seem to have missed that part (and changed the docs too eagerly). 

Thanks for the find! I'll revert the changes for all three affected sections :+1:

File: packages/cli/index.js (hunk 3 above)
Comment (lhw-1) [reply packages/cli/index.js:102]: Actually nope, since the `dev` mode is already available, this is quite redundant :sweat_smile: Let me revert this!

Comment (lhw-1) [reply packages/cli/index.js]: Fixed in [a27bad9](https://github.com/MarkBind/markbind/pull/2218/commits/a27bad98b61f596abe72f9dfc46d580724ddd2cf)!

File: docs/userGuide/deployingTheSite.md [hunk 4]
```diff
@@ -23,17 +23,28 @@
...
 1. Upload the site files to the Web server. The sections below explain how to automate this step if you are deploying to some online platforms.
 
 **Steps for deploying multiple MarkBind sites:**
 
 1. Create multiple `site.json` files. Ensure that the [`baseUrl` property of each `site.json` file](siteJsonFile.html#baseurl) matches its deploy location.
 1. (Optional) Use the [`markbind serve -s <file>` command](cliCommands.html#serve-command) to stage each site locally and confirm the contents are as expected.
 1. For each site:
-    1. Use the [`markbind build -s <file>` command](cliCommands.html#build-command) to generate the site from source files.
+    1. (Optional) Use the [`markbind build -s <file>` command](cliCommands.html#build-command) to generate the site from source files. Note that the [`markbind deploy` command](cliCommands.html#deploy-command) generates the site from source files by default as well.
```
[the remark this answers is not in the document]
Comment (lhw-1) [reply docs/userGuide/deployingTheSite.md]: Fixed in [a27bad9](https://github.com/MarkBind/markbind/pull/2218/commits/a27bad98b61f596abe72f9dfc46d580724ddd2cf)!

File: docs/userGuide/deployingTheSite.md (hunk 2 above)
Comment (lhw-1) [reply docs/userGuide/deployingTheSite.md]: Fixed in [a27bad9](https://github.com/MarkBind/markbind/pull/2218/commits/a27bad98b61f596abe72f9dfc46d580724ddd2cf)!

File: docs/userGuide/addingPages.md (hunk 1 above)
Comment (lhw-1) [reply docs/userGuide/addingPages.md]: Fixed in [a27bad9](https://github.com/MarkBind/markbind/pull/2218/commits/a27bad98b61f596abe72f9dfc46d580724ddd2cf)!

**Review by jovyntls — COMMENTED** [review]
LGTM, mainly nits on documentation!
Remember to use the commit message format as well.

File: docs/userGuide/cliCommands.md [hunk 5]
```diff
@@ -174,22 +174,40 @@ The caveat is that not building all pages during the initial process, or not reb
...
 * `-c <githubTokenName>`, `--ci <githubTokenName>`<br>
-  Deploy the site in CI Environments using the GitHub personal access token stored in `<githubTokenName>`. (default: `GITHUB_TOKEN`)<br>
+  Deploy the site in CI Environments using the GitHub personal access token stored in `<githubTokenName>` (default: `GITHUB_TOKEN`).<br>
   {{ icon_example }} `-c PA_TOKEN`
 
+* `-n`, `--no-build`<br>
+  Skips the generation of the site before deploying it. By default, the `deploy` command generates the site before deploying the built site. This command should be used if the site has been generated with specific arguments (e.g. baseUrl is specified during the `markbind build` command).
+
+* `-s <file>`, `--site-config <file>`<br>
+  Specify the site config file (default: `site.json`)<br>
```
[the remark this answers is not in the document]
Comment (lhw-1) [reply docs/userGuide/cliCommands.md]: Updated in [9576afb](https://github.com/MarkBind/markbind/pull/2218/commits/9576afb0e4210a56b8db84500feb482cb7c62fc2)!

File: docs/userGuide/cliCommands.md [hunk 6]
```diff
@@ -174,22 +174,40 @@ The caveat is that not building all pages during the initial process, or not reb
...
+  {{ icon_example }} `-s otherSite.json`
+
 %%{{ icon_info }} Related: [User Guide: Deploying the Website](deployingTheSite.html).%%
 
+**{{ icon_examples }}**
+* `markbind deploy`
+* `markbind deploy ./myWebsite` : Generates and deploys the site from the `myWebsite` directory.
+* `markbind deploy --no-build` : Deploys the site without generating it first.
+directory.
```
[the remark this answers is not in the document]
Comment (lhw-1) [reply docs/userGuide/cliCommands.md]: Updated in [9576afb](https://github.com/MarkBind/markbind/pull/2218/commits/9576afb0e4210a56b8db84500feb482cb7c62fc2)!

File: docs/userGuide/cliCommands.md [hunk 7]
```diff
@@ -174,22 +174,40 @@ The caveat is that not building all pages during the initial process, or not reb
...
 
 **Options** :fas-cogs:
 
 * `-c <githubTokenName>`, `--ci <githubTokenName>`<br>
-  Deploy the site in CI Environments using the GitHub personal access token stored in `<githubTokenName>`. (default: `GITHUB_TOKEN`)<br>
+  Deploy the site in CI Environments using the GitHub personal access token stored in `<githubTokenName>` (default: `GITHUB_TOKEN`).<br>
   {{ icon_example }} `-c PA_TOKEN`
 
+* `-n`, `--no-build`<br>
+  Skips the generation of the site before deploying it. By default, the `deploy` command generates the site before deploying the built site. This command should be used if the site has been generated with specific arguments (e.g. baseUrl is specified during the `markbind build` command).
```
[the remark this answers is not in the document]
Comment (lhw-1) [reply docs/userGuide/cliCommands.md]: In [9576afb](https://github.com/MarkBind/markbind/pull/2218/commits/9576afb0e4210a56b8db84500feb482cb7c62fc2), I reworded this as:

"This command should be used if the site needs to be generated with non-default arguments for the `build` command (e.g. baseUrl is specified using the `--baseUrl` flag for the `markbind build` command)."

Does this sound better (or perhaps equally confusing)? :sweat_smile:

File: docs/userGuide/deployingTheSite.md [hunk 8]
```diff
@@ -66,10 +66,22 @@ You can override the default deployment settings %%(e.g., repo/branch to deploy)
 
 <br>
 
+<box type="info">
+
+`markbind deploy` generates the static site from your source by default; hence, it is not necessary to run `markbind build` first to generate the site before deploying.
+
+</box>
+
+<div id="warning-about-build-and-deploy">
```
[the remark this answers is not in the document]
Comment (lhw-1) [reply docs/userGuide/deployingTheSite.md]: Ah, not really at the moment. Removed in [9576afb](https://github.com/MarkBind/markbind/pull/2218/commits/9576afb0e4210a56b8db84500feb482cb7c62fc2)!

File: docs/userGuide/deployingTheSite.md [hunk 9]
```diff
@@ -66,10 +66,22 @@ You can override the default deployment settings %%(e.g., repo/branch to deploy)
...
+
 <box type="warning">
 
-`markbind deploy` does not generate the static site from your source; it simply deploys the files that are already in the `_site` directory. You need to run `markbind build` first if you want to generate the site before deploying.
+If you have generated the site with specific arguments, e.g. `markbind build --baseUrl staging`, or wish to deploy without generating the static site first for any reason, you should deploy without the automatic generation, i.e. by using `markbind deploy --no-build`. This will allow the deployment of the site to happen from the previously generated site without accidentally re-generating it from the wrong directory.
```
[the remark this answers is not in the document]
Comment (lhw-1) [reply docs/userGuide/deployingTheSite.md]: That's a good point - in [9576afb](https://github.com/MarkBind/markbind/pull/2218/commits/9576afb0e4210a56b8db84500feb482cb7c62fc2), I've combined the second and third boxes to make it more readable!

**Review by jovyntls — APPROVED** [review]
LGTM, thanks for the docs additions & detailed commit message @lhw-1 !

When it's ready (by the DG project management guidelines), feel free to try merging this PR! Remember to use the squash-merge strategy with the commit message :)

**Pull request merged**